modules/ipp/scanner.go

```diff
@@ -225,80 +211,149 @@ func bufferFromBody(res *http.Response, scanner *Scanner) *bytes.Buffer {
 }
 
 // FIXME: This will read the wrong section of the body if a substring matches the attribute name passed in
-// TODO: Support reading from multiple instances of the same attribute in a response
-func readAttributeFromBody(attrString string, body *[]byte) ([][]byte, error) {
+func readAttributeFromBody(attrString string, body *[]byte) ([][][]byte, error) {
 	attr := []byte(attrString)
 	interims := bytes.Split(*body, attr)
-	if len(interims) > 1 {
-		valueTag := interims[0][len(interims[0])-3]
+	var valSlice [][][]byte
+	if len(interims) <= 1 {
+		//The attribute was not present
+		return nil, errors.New("Attribute \"" + attrString + "\" not present.")
+	}
+	for i, interim := range interims[:len(interims)-1] {
+		valueTag := interim[len(interim)-3]
 		var vals [][]byte
-		buf := bytes.NewBuffer(interims[1])
+		buf := bytes.NewBuffer(interims[i+1])
 		// This reading occurs in a loop because some attributes can have type "1 setOf <type>"
 		// where same attribute has a set of values, rather than one
 		for tag, nameLength := valueTag, int16(0); tag == valueTag && nameLength == 0; {
 			var length int16
 			if err := binary.Read(buf, binary.BigEndian, &length); err != nil {
 				//Couldn't read length of content
-				return vals, err
+				break
 			}
 			val := make([]byte, length)
 			if err := binary.Read(buf, binary.BigEndian, &val); err != nil {
 				//Couldn't read content
-				vals = append(vals, val)
-				return vals, err
+				//vals = append(vals, val)
+				break
 			}
 			vals = append(vals, val)
 			if err := binary.Read(buf, binary.BigEndian, &tag); err != nil {
 				//Couldn't read next valueTag
-				return vals, err
+				break
 			}
 			// FIXME: Only try to read next namelength if previous valueTag wasn't end-of-attributes-tag
 			if err := binary.Read(buf, binary.BigEndian, &nameLength); err != nil {
 				//Couldn't read next nameLength
-				return vals, err
+				break
 			}
 		}
-		return vals, nil
+		valSlice = append(valSlice, vals)
 	}
-	//The attribute was not present
-	return nil, errors.New("Attribute \"" + attrString + "\" not present.")
+	return valSlice, nil
 }
 
-func (scan *scan) tryReadAttributes(body string) {
-	bodyBytes := []byte(body)
-	// Write reported CUPS version to results object
-	if scan.results.AttributeCUPSVersion == "" {
-		if cupsVersions, err := readAttributeFromBody(CupsVersion, &bodyBytes); err != nil {
-			log.WithFields(log.Fields{
-				"error":     err,
-				"attribute": CupsVersion,
-			}).Debug("Failed to read attribute.")
-		} else if len(cupsVersions) > 0 {
-			scan.results.AttributeCUPSVersion = string(cupsVersions[0])
+type Value struct {
+	Bytes []byte `json:"raw,omitempty"`
+}
+
+type Attribute struct {
+	Name string    `json:"name,omitempty"`
+	Values []Value `json:"values,omitempty"`
+	ValueTag byte  `json:"tag,omitempty"`
+}
+
+type AttributeGroup struct {
+	Attrs []*Attribute
+	Tag byte
+}
+
+func printGroup(g *AttributeGroup) {
```

Comment:
Unused?

Reply:
Yeah, I was using that to troubleshoot a bit while writing the code. I'll remove it.